drivers/ethernet/dwc_xgmac/eth_dwc_xgmac.c

```diff
@@ -0,0 +1,1943 @@
+#define UPDATE_ETH_STATS_TX_BYTE_CNT(dev_data, incr)
+#define UPDATE_ETH_STATS_RX_BYTE_CNT(dev_data, incr)
+#define UPDATE_ETH_STATS_TX_ERROR_PKT_CNT(dev_data, incr)
+#define UPDATE_ETH_STATS_RX_ERROR_PKT_CNT(dev_data, incr)
+#define UPDATE_ETH_STATS_TX_DROP_PKT_CNT(dev_data, incr)
+#endif
+
+/**
+ * @brief Run-time device configuration data structure.
+ *
+ * This struct contains all device configuration data for a XGMAC
+ * controller instance which is modifiable at run-time, such as
+ * data relating to the attached PHY or the auxiliary thread.
+ */
+struct eth_dwc_xgmac_dev_data {
+	DEVICE_MMIO_RAM;
+	/*
+	 * Device running status. eth_dwc_xgmac_start_device API will will variable and
+	 * eth_dwc_xgmac_stop_device APi will clear this variable.
+	 */
+	bool dev_started;
+	/* Ethernet interface structure associated with this device. */
+	struct net_if *iface;
+	/* Device MAC address */
+	uint8_t mac_addr[6];
+	/* Current Ethernet link speed 10Mbps/100Mbps/1Gbps. */
+	enum eth_dwc_xgmac_link_speed link_speed;
+	/* This field specifies the ethernet link type full duplex or half duplex. */
+	bool enable_full_duplex;
+	/* Ethernet auto-negotiation status. */
+	bool auto_neg;
+	/* Ethernet promiscuous mode status. */
+	bool promisc_mode;
+	/* Global pointer to DMA receive descriptors ring. */
+	struct xgmac_dma_rx_desc *dma_rx_desc;
+	/* Global pointer to DMA transmit descriptors ring. */
+	struct xgmac_dma_tx_desc *dma_tx_desc;
+	/* Global pointer to DMA transmit descriptors ring's meta data. */
+	volatile struct xgmac_dma_tx_desc_meta *tx_desc_meta;
+	/* Global pointer to DMA receive descriptors ring's meta data. */
+	volatile struct xgmac_dma_rx_desc_meta *rx_desc_meta;
+	/*
+	 * Array of pointers pointing to Transmit packets  under transmission.
+	 * These pointers will be cleared once the packet transmission is completed.
+	 */
+	mem_addr_t *tx_pkts;
+	/*
+	 * Array of pointers pointing to receive fragments reserved for receive data.
+	 * Data received by XGMAC will be copied to these fragments. An empty network packet
+	 * will reserved as receive packet, these fragments will be added to a the receive packet.
+	 * After a fragment is added to the receive packet a new fragment will be reserved
+	 * and replaced with the used fragment for future receive data.
+	 */
+	mem_addr_t *rx_frags;
+	/* A global pointer pointing to XGMAC IRQ context data. */
+	struct xgmac_irq_cntxt_data irq_cntxt_data;
+#ifdef CONFIG_NET_STATISTICS_ETHERNET
+	/* Ethernet statistics captured by XGMAC driver */
+	struct net_stats_eth stats;
+#endif /* CONFIG_NET_STATISTICS_ETHERNET */
+#ifdef CONFIG_ETH_DWC_XGMAC_POLLING_MODE
+	/* timer for interrupt polling */
+	struct k_timer isr_polling_timer;
+#endif /* CONFIG_ETH_DWC_XGMAC_POLLING_MODE */
+#ifdef CONFIG_ETH_DWC_XGMAC_BOTTOM_HALF_WORK_QUEUE
+	/* work que item for processing TX interrupt bottom half */
+	struct k_work isr_work;
+#endif /*CONFIG_ETH_DWC_XGMAC_BOTTOM_HALF_WORK_QUEUE*/
+	struct k_mutex dev_cfg_lock;
+};
```

Comment:
check structure alignment, rom/ram scarcity is probably not an issue in your case but it's a good habit to follow in general. (for instance dev_started could be moved with other bools below, mac_addr creates a gap of 2 bytes as well so up to you to find a better place, if not at the end etc etc...).

Reply:
Reordered the structure members for better alignment.